File: combinationSumForLoopRecursion.java
class Solution {
    List<List<Integer>> result;
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        result = new ArrayList<>();
        if(candidates == null || candidates.length == 0) {
            return result;
        }
        helper(candidates, target, 0, new ArrayList<>());
        return result;
    }
    private void helper(int[] candidates, int target, int index, List<Integer> path) {
        //base
        if(target == 0) {
            result.add(path);
            return;
        }
        if(target < 0) {
            return;
        }
       
        for(int i = index; i < candidates.length; i++) {
            List<Integer> list = new ArrayList<>(path);
            list.add(candidates[i]);
            helper(candidates, target - candidates[i], i, list);
        }
    }
}

File: combination_sum_for_loop_recursion.py
"""
Time Complexity : O(2^m*n) where m is the length of candidates and n is the target
Space Complexity : O(m+n)  where m is the length of candidates and n is the target

Did this code successfully run on Leetcode : Yes
Any problem you faced while coding this : No
"""


class Solution:
    def __init__(self):
        self.result = []

    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        if len(candidates) == 0:
            return self.result

        self.helper(candidates, target, 0, [])
        return self.result

    def helper(self, candidates, amount, pivot, path):
        # base case
        if amount == 0:
            self.result.append(path[:])
            return
        if amount < 0:
            return

        # logic
        for i in range(pivot, len(candidates)):
            path.append(candidates[i])
            self.helper(candidates, amount - candidates[i], i, path)
            path.pop()
